### PR TITLE
sriov CNI, Allocate only one sriov PF

### PR DESCRIPTION
**Pull request by oshoval**
In order to allow running multi sriov clusters simultaneously,
allocate only one sriov PF per CNI call.
    
In case no PFs exist, exit with an error, so the job won't
start without PF.
It may occur in case the container was forcely deleted,
and then it takes a few minutes for the PF
to reappear in `/sys/class/net`.
The CNI will exit, delete the old NS, and a new container will
be created and retrigger CNI that would try again, until
the PF reappears.

This update makes the CNI re-entrant as well,
by detecting possible collisions in case two jobs try to take the same PF
at the same time.
One of them will succeed, and the other will advance to the next available PF.
Prow will schedule jobs according the number of PFs on the node (will be done on other PR).
This way we are guaranteed to be safe.

The CNI doesn't mess with VFs at all, kubevirtci rule is to do so.

Additional changes:
* Remove unneeded conf file.
    The NAD configuration that is used is the
    `NetworkAttachmentDefinition`, not the conf file one.
    Hence remove it.
* Remove unneeded veth code.

Note:
In order to allocate more than one PF for job, call this CNI twice.

CNI container image: `quay.io/kubevirtci/sriov-passthrough-cni:v1.1`

Signed-off-by: Or Shoval <oshoval@redhat.com>

### CODE REVIEW

File: cni-plugins/sriov-passthrough-cni/deploy/multus-cni.yaml
Comment (oshoval) [inline cni-plugins/sriov-passthrough-cni/deploy/multus-cni.yaml:160]: Once there is an ACK on the PR, we can push it to an official repo and change just this yaml on CI,
in order to take those changes.

File: cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni
Comment (qinqon) [inline cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni:30]: We have to block this function with a mutex or the like on the node.

File: cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni
Comment (qinqon) [inline cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni:44]: Are we going to need more than one PF per job ? if not we can simplify all this and remove PF_COUNT

**Review by qinqon — CHANGES_REQUESTED** [review]

File: cni-plugins/sriov-passthrough-cni/deploy/multus-cni.yaml
Comment (oshoval) [reply cni-plugins/sriov-passthrough-cni/deploy/multus-cni.yaml:159]: It just need the NAD right, i tested it on CI (its deployed there actually now with a different name), it doesn't have the conf and it works perfectly, using the NAD (made sure that as well, when the NAD was on wrong namespace it didn't work).
Based on review #479

File: cni-plugins/sriov-passthrough-cni/deploy/multus-cni.yaml
Comment (oshoval) [reply cni-plugins/sriov-passthrough-cni/deploy/multus-cni.yaml:160]: yep will do, thanks

File: cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni
Comment (oshoval) [reply cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni:30]: we don't need, its re-entrant because a PF can exists only on one NS, so it checks if it managed to move the PF,
if it didn't it continue to the next one.

File: cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni
Comment (oshoval) [reply cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni:44]: For migration we might need to move 2, but we can do it with calling the CNI twice (see PR description note please).
We can remove this, but i think it doesn't hurt, because if we will want to make it static 2, instead calling the CNI twice, it will allow to just change PF_COUNT, wdyt ?

File: cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni
Comment (qinqon) [reply cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni:30]: Two jobs running at the same time can end up with the same sriov_pfs list so one of them will fail at ip link set and the job will fail, if we use a mutex at node somehow one of them will just wait a little.

File: cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni
Comment (qinqon) [reply cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni:44]: Since we cannot pass PF_COUNT I don't know how migration test will change it when CNI is called, but also I don't know what is needed to call this cni twi for a specific test.

File: cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni
Comment (oshoval) [reply cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni:30]: The list will be the same, but It won't fail, because there are more PFs which will be enough for each schedule job,
it will just detect the collision and allocate the next one.
We won't spin more than available amount of PFs
see please https://github.com/kubevirt/project-infra/pull/784
This is the design
Also mentioned it in the PR description.

File: cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni
Comment (oshoval) [reply cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni:44]: Migration test won't change it, its either that we make it static, or we call two annotations from the job yaml.
It's just to support N allocation, if needed i will remove it, i just don't think it hurts.

File: cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni
Comment (qinqon) [reply cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni:30]: Ahh right it will go to the next PF and that's all.

File: cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni
Comment (qinqon) [reply cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni:44]: Ok, good enough.

File: cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni
Comment (qinqon) [inline cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni:42]: Little nit, can we make an if statement out of this ? feels super weird to do a log with &&

File: cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni
Comment (qinqon) [inline cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni:33]: Same here let's add a if statement

File: cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni
Comment (oshoval) [reply cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni:42]: I am +1 for that, Edy didn't like that style on other PR
@EdDev  ?

Comment (qinqon) [inline cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni:42]: ditto, better not inline this.

Comment (oshoval) [reply cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni:42]: I am +1 for that, @EdDev didn't like `if`, @EdDev is it fine by you that i change it to if ?

File: cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni
Comment (oshoval) [reply cni-plugins/sriov-passthrough-cni/plugin/sriov-passthrough-cni:33]: lets see what the result of https://github.com/kubevirt/project-infra/pull/884#discussion_r565317084
and do accordingly

**Review by qinqon — CHANGES_REQUESTED** [review]
All good just a pair of nits.

File: cni-plugins/sriov-passthrough-cni/deploy/multus-cni.yaml
Comment (oshoval) [reply cni-plugins/sriov-passthrough-cni/deploy/multus-cni.yaml:160]: once the PR will be done, we can push it to 
https://quay.io/repository/kubevirtci/sriov-passthrough-cni
thanks @fgimenez

**Comment by oshoval** [comment]
Addressed comments, updated container image (need to build and push it there once this PR is lgtmed).

**Comment by qinqon** [comment]
@oshoval are we going to add some docs about how to use it at prow jobs ?

**Comment by oshoval** [comment]
> @oshoval are we going to add some docs about how to use it at prow jobs ?

Yes, if possible lets not block this PR, i am preparing it atm.

File: cni-plugins/sriov-passthrough-cni/deploy/multus-cni.yaml
Comment (oshoval) [reply cni-plugins/sriov-passthrough-cni/deploy/multus-cni.yaml]: Imaged was built and pushed by Federico, thanks.

**Comment by qinqon** [comment]
@oshoval I am not approver/owner of project-infra anymore (https://github.com/kubevirt/project-infra/pull/885) so I can only LGTM.

**Comment by oshoval** [comment]
Thanks Quique, will deploy it manually and then ask a maintainer to merge it.

**Comment by EdDev** [comment]
> Thanks Quique, will deploy it manually and then ask a maintainer to merge it.

It must be merged first. Otherwise we may have a sync problem.

**Comment by oshoval** [comment]
/cc @fgimenez 
Would you take a look please?

**Comment by fgimenez** [comment]
@oshoval sure, looks good

/approve

**Pull request merged**
